src/bitmap/store/array_store/visitor.rs

```diff
@@ -0,0 +1,32 @@
+pub trait BinaryOperationVisitor {
+    fn visit_scalar(&mut self, value: u16);
+    fn visit_slice(&mut self, values: &[u16]);
```

Comment:
Could you maybe add a small documentation line here? Just because `visit` is not quite clear when the impl of `VecWriter` is to push in an internal Vec.

Reply:
In fact if it does make more sense when you'll introduce SIMD, it's maybe not that interesting to document RN.

Reply:
Out of scope for this PR but here is justification for the visitor abstraction:

```rust
pub struct CardinalityCounter {
    count: usize,
}

impl CardinalityCounter {
    pub fn new() -> CardinalityCounter {
        CardinalityCounter { count: 0 }
    }

    pub fn into_inner(self) -> usize {
        self.count
    }
}

impl ArrayBinaryOperationVisitor for CardinalityCounter {
    fn visit_vector(&mut self, _value: u16x8, mask: u8) {
        self.count += mask.count_ones() as usize;
    }

    fn visit_scalar(&mut self, _value: u16) {
        self.count += 1;
    }

    fn visit_slice(&mut self, values: &[u16]) {
        self.count += values.len();
    }
}
```

Compared to the [C implementation](https://github.com/RoaringBitmap/CRoaring/blob/master/src/array_util.c#L450)

In Rust the algos monomorphize and we get the same result without having to write everything twice.

Reply:
You can pass in that visitor to any `scalar::and` `vector::and`, `scalar::or` .. and so on. This can also be used to implement Jaccard index, which might be applicable to meli?

Reply:
Does the Jaccard index give a number between 0 and 1 which indicates the similarity of two sets? Not sure where it can be used in milli, but maybe somewhere, I didn't check that.

Reply:
@Kerollmops It could be used for any metric: Jaccard, cosine similarity, etc. Any time you'd want the cardinality of some ops without having to create or mutate a bitmap, this will be quite a bit faster

Reply:
This is awesome.